libraries/stdlib/osl/stdosl.h

```diff
@@ -573,6 +573,169 @@ closure color volume_matte() BUILTIN;
+
+// Constructs a closure that represents straight transmission through a surface.
+//
+// NOTE:
+//  - This is not a node in the MaterialX library, but the surface shader constructor
+//    node has an 'opacity' parameter to control textured cutout opacity.
+//
+closure color transparent_bsdf() BUILTIN;
+​
+// Constructs a BSSRDF for subsurface scattering within a homogeneous medium.
+//
+// TODO:
+// - Do we want/need this dedicated BSSRDF closure? We have had discussions before about
+//   describing this more rigorously as a BSDF layered over a VDF instead.
+//   In the MaterialX library we want this as a node for the user friendly interface,
+//   but for back-end implementation maybe a dedicated closure is not needed?
+//
+closure color subsurface_bssrdf(normal N, color albedo, color radius, float anisotropy) BUILTIN;
+​
+// Constructs a microfacet BSDF for the back-scattering properties of cloth-like materials.
+// This closure may be vertically layered over a base BSDF, where energy that is not reflected
+// will be transmitted to the base closure.
+//
+closure color sheen_bsdf(normal N, color albedo, float roughness) BUILTIN;
+​
+// Adds an iridescent thin film layer over a microfacet base BSDF. This must be layered over
+// another base BSDF using the layer() closure, as this is a modifier and cannot be used as a
+// standalone closure.
+//
+// TODO:
+// - This might be better to represent as optional extra arguments on the closures that do support
+//   thin-film iridescence (dielectric_bsdf, conductor_bsdf and generalized_schlick_bsdf)?
+//
+closure color thin_film_bsdf(float thickness, float ior) BUILTIN;
+​
+​
+// -------------------------------------------------------------//
+// EDF closures                                                 //
+// -------------------------------------------------------------//
+​
+// Constructs an EDF emitting light uniformly in all directions.
+//
+closure color uniform_edf(color emittance) BUILTIN;
+
+// Constructs an EDF emitting light inside a cone around the normal direction.
+//
+closure color conical_edf(color emittance, normal N, float inner_angle, float outer_angle) BUILTIN;
+
+// Constructs an EDF emitting light according to a measured IES light profile.
+//
+closure color measured_edf(color emittance, normal N, string file) BUILTIN;
+​
+​
+// -------------------------------------------------------------//
+// VDF closures                                                 //
+// -------------------------------------------------------------//
+​
+// Constructs a VDF for pure volumetric absorption.
+//
+closure color absorption_vdf(color absorption) BUILTIN;
+​
+// Constructs a VDF scattering light for a participating medium, based on the Henyey-Greenstein
+// phase function. Forward, backward and uniform scattering is supported and controlled by the
+// anisotropy input.
+//
+closure color anisotropic_vdf(color absorption, color scattering, float anisotropy) BUILTIN;
```

Comment:
absorption_vdf seems redundant when you have this. Also, absorption_vdf is not really a VDF strictly speaking. Should we just have one VDF like anisotropic_vdf? Maybe with (color attenuation, color albedo, ...) parametrization? You can always pass 0 albedo if you just want absorption.

Question: should we expect the params to be premultiplied by density? I would think yes.

At SPI we also chose to have a medium closure that sets all params for the medium enclosed by the current interface (when shading surfaces). It is very useful for glass/liquid objects. Do we want that?

Reply:
Yep, the absorption_vdf if just a subset of anisotropic_vdf and it can be removed unless there is a performance benefit of having this as a special case. But I think you are right we should just remove it.

Your proposal sounds interesting, maybe you could write it out to clarify the full parameterization. Same thing for the special medium closure. And I hope others will chime in as well :)

Reply:
Sure, this is our medium closure prototype:

closure color medium(color sigma_a, color sigma_s,  float anisotropy, float ior, int priority) BUILTIN;

And it is designed so you return it together with say, dielectric:

Ci = dielectric(N, ...) + medium(sigma_a, ...)

and the semantics is that you are telling the render this is a closed object with that interior medium. You can then use medium tracking with priorities to override relative IOR at the interface if your object is say, underwater. Note that the IOR of the medium is absolute, while the one on the surface interface is relative.

In theory you could throw away the ior param to the BSDFs and drive it entirely from the medium tracking. We haven't done so as most of our shaders don't define a medium, we just override the one in the BSDF when the medium is present.

Reply:
Arnold also has a (dielectric) medium closure, which is basically the same as Alejandro described. I think it could be useful to have this explicit in the model here. As discussed in Slack, one could actually replace the SSS with the concept of the interior medium, and map the artist-friendly SSS parameters to the medium parameters.

Though I'm not sure about the adding of dielectric and medium closures denoting the combination of the BSDF and the underlying medium. The BSDF and its surrounding media should ideally be tied together in some explicit way.

Reply:
I agree that using addition for combining them seems a bit unintuitive. Using the layer() closure for this as well might be more logical conceptually:
`Ci = layer(dielectric(N, ...), medium(sigma_s, ...))`
Stating to the renderer we are placing a medium underneath the BSDF interface.

Reply:
About the `medium` closure itself, you are saying the implementation for this would be different from the `anisotropic_vdf` closure, right? With the special handling of IOR and nested priority? So it would make sense to have _both_ the anisotropic_vdf closure and this medium closure as part of the standard closures? If so we should consider to make this medium closure a dedicated MaterialX nodes as well.

Reply:
I would think of the `medium` mostly just as some container specifying the (homogeneous) properties of the bulk, i.e.:

- IOR
- priority (optional?)
- dispersion (optional?)
- volume/SSS properties

Where the volume properties are:

- sigma_a, sigma_s   (or equivalently, extinction and albedo)
- phase_function

The only distribution function ("VDF") involved here is the phase function, which is usually parametrized by a single anisotropy value. Alternatively, one could just specify the anisotropy value with the phase function some hard-coded form (probably Henyey-Greenstein) or implementation specific.

Reply:
Or we could merge everything in a single vdf closure that would add IOR and priority to the volume properties. If you are shading an actual volume those would be unused (although somebody may want to go fancy and bend rays with IOR in the air). The homogeneous/heterogeneous character would come from the context where the closure is evaluated. Unless you guys feel it is important for that distinction to be explicit.

Btw, I'm leaning towards extinction/albedo parametrization for volume. I think we always end up computing those for rendering (sigma_t/albedo).

Reply:
@portsmouth @aconty Is this new parameterization of anisotropic_vdf matching what you had in mind? I also added ior & priority as documented optional parameters for homogeneous media.